modules/core/04-channel/migrations/v10/store.go

```diff
@@ -0,0 +1,117 @@
+package v10
+
+import (
+	fmt "fmt"
+
+	corestore "cosmossdk.io/core/store"
+	storetypes "cosmossdk.io/store/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/runtime"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/ibc-go/v10/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/v10/modules/core/24-host"
+)
+
+const (
+	// ParamsKey defines the key to store the params in the keeper.
+	ParamsKey               = "channelParams"
+	KeyPruningSequenceStart = "pruningSequenceStart"
+
+	KeyChannelUpgradePrefix = "channelUpgrades"
+	KeyUpgradePrefix        = "upgrades"
+	KeyUpgradeErrorPrefix   = "upgradeError"
+	KeyCounterpartyUpgrade  = "counterpartyUpgrade"
+)
+
+// PruningSequenceStartKey returns the store key for the pruning sequence start of a particular channel
+func PruningSequenceStartKey(portID, channelID string) []byte {
+	return []byte(fmt.Sprintf("%s/%s", KeyPruningSequenceStart, channelPath(portID, channelID)))
+}
+
+func ChannelUpgradeKey(portID, channelID string) []byte {
+	return []byte(fmt.Sprintf("%s/%s/%s", KeyChannelUpgradePrefix, KeyUpgradePrefix, channelPath(portID, channelID)))
+}
+
+func ChannelUpgradeErrorKey(portID, channelID string) []byte {
+	return []byte(fmt.Sprintf("%s/%s/%s", KeyChannelUpgradePrefix, KeyUpgradeErrorPrefix, channelPath(portID, channelID)))
+}
+
+func ChannelCounterpartyUpgradeKey(portID, channelID string) []byte {
+	return []byte(fmt.Sprintf("%s/%s/%s", KeyChannelUpgradePrefix, KeyCounterpartyUpgrade, channelPath(portID, channelID)))
+}
+
+func channelPath(portID, channelID string) string {
+	return fmt.Sprintf("%s/%s/%s/%s", host.KeyPortPrefix, portID, host.KeyChannelPrefix, channelID)
+}
+
+func MigrateStore(ctx sdk.Context, storeService corestore.KVStoreService, cdc codec.BinaryCodec, channelKeeper ChannelKeeper) error {
+	store := storeService.OpenKVStore(ctx)
+
+	if err := handleChannelMigration(ctx, store, cdc, channelKeeper); err != nil {
+		return err
+	}
+	if err := deleteChannelUpgrades(store); err != nil {
+		return err
+	}
+	if err := deleteParams(store); err != nil {
+		return err
+	}
+	if err := deletePruneSequences(store); err != nil {
+		return err
+	}
+
+	// TODO: See if there is more to migrate/delete from store
+
+	return nil
+}
+
+func handleChannelMigration(ctx sdk.Context, store corestore.KVStore, cdc codec.BinaryCodec, channelKeeper ChannelKeeper) error {
+	// Remove channel upgrade sequences and set in-upgrade channels back to open
+	iterator := storetypes.KVStorePrefixIterator(runtime.KVStoreAdapter(store), []byte(host.KeyChannelEndPrefix))
+
+	defer sdk.LogDeferred(channelKeeper.Logger(ctx), func() error { return iterator.Close() })
+	for ; iterator.Valid(); iterator.Next() {
+		var channel Channel
+		cdc.MustUnmarshal(iterator.Value(), &channel)
+
+		if channel.State == FLUSHING || channel.State == FLUSHCOMPLETE {
```

Comment:
This will mess with the channel upgrade procedure. If we have a FLUSHCOMPLETE on one side, we are expecting the upgrade to occur successfully.

So if you downgrade from this point, you risk the other side upgrading successfully using your FLUSHCOMPLETE state and having the channel be invalid.

I would suggest that we just panic migrating this if any channel is in FLUSHING or FLUSHCOMPLETE.

Since these are gated by governance ultimately, we can do the upgrade to v10 when we have no active upgrades in-flight.

Creates much less headache for us thinking through these cases

Reply:
Yeah, that makes sense. I was considering that, but figured it might just work out if both sides migrated. I guess panic might be safer in this case. Will fix.

Reply:
It will error instead now